src/LanguageServer/Protocol/Handler/Diagnostics/DiagnosticsPullCache.cs

```diff
@@ -0,0 +1,123 @@
+    {
+        public override async Task<(int globalStateVersion, VersionStamp? dependentVersion)> ComputeCheapVersionAsync(DiagnosticsRequestState state, CancellationToken cancellationToken)
+        {
+            return (state.GlobalStateVersion, await state.Project.GetDependentVersionAsync(cancellationToken).ConfigureAwait(false));
+        }
+
+        public override async Task<(int globalStateVersion, Checksum dependentChecksum)> ComputeExpensiveVersionAsync(DiagnosticsRequestState state, CancellationToken cancellationToken)
+        {
+            return (state.GlobalStateVersion, await state.Project.GetDependentChecksumAsync(cancellationToken).ConfigureAwait(false));
+        }
+
+        public override async Task<ImmutableArray<DiagnosticData>> ComputeDataAsync(DiagnosticsRequestState state, CancellationToken cancellationToken)
+        {
+            var diagnostics = await state.DiagnosticSource.GetDiagnosticsAsync(state.Context, cancellationToken).ConfigureAwait(false);
+            state.Context.TraceInformation($"Found {diagnostics.Length} diagnostics for {state.DiagnosticSource.ToDisplayString()}");
+            return diagnostics;
+        }
+
+        public override Checksum ComputeChecksum(ImmutableArray<DiagnosticData> data)
+        {
+            // Create checksums of diagnostic data and sort to ensure stable ordering for comparison.
+            var diagnosticDataChecksums = data
+                .SelectAsArray(d => Checksum.Create(d, SerializeDiagnosticData))
+                .Sort();
+
+            return Checksum.Create(diagnosticDataChecksums);
+        }
+
+        private static void SerializeDiagnosticData(DiagnosticData diagnosticData, ObjectWriter writer)
+        {
+            writer.WriteString(diagnosticData.Id);
+            writer.WriteString(diagnosticData.Category);
+            writer.WriteString(diagnosticData.Message);
+            writer.WriteInt32((int)diagnosticData.Severity);
+            writer.WriteInt32((int)diagnosticData.DefaultSeverity);
+            writer.WriteBoolean(diagnosticData.IsEnabledByDefault);
+            writer.WriteInt32(diagnosticData.WarningLevel);
+
+            // Ensure the tag order is stable before we write it.
+            foreach (var tag in diagnosticData.CustomTags.Sort())
+            {
+                writer.WriteString(tag);
+            }
+
+            foreach (var key in diagnosticData.Properties.Keys.ToImmutableArray().Sort())
+            {
+                writer.WriteString(key);
+                writer.WriteString(diagnosticData.Properties[key]);
+            }
+
+            if (diagnosticData.ProjectId != null)
+                writer.WriteGuid(diagnosticData.ProjectId.Id);
+
+            WriteDiagnosticDataLocation(diagnosticData.DataLocation, writer);
+
+            foreach (var additionalLocation in diagnosticData.AdditionalLocations)
```

Comment:
>AdditionalLocations [](http://example.com/codeflow?start=62&length=19)

Are there usually very many items in here? If so, do they usually have the same file path?

Reply:
Oh, I didn't notice there were multiple file spans per location. Maybe these can't be easily grouped by file location.